src/frontend/src/catalog/index_catalog.rs

```diff
@@ -0,0 +1,151 @@
+// Copyright 2022 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use itertools::Itertools;
+use risingwave_common::catalog::IndexId;
+use risingwave_common::types::DataType;
+use risingwave_pb::catalog::Index as ProstIndex;
+use risingwave_pb::expr::expr_node::RexNode;
+
+use crate::catalog::{DatabaseId, SchemaId, TableId};
+use crate::expr::{Expr, InputRef};
+
+/// # `IndexCatalog` Example:
+/// create table t (a int, b int, c int, d int);
+/// create index idx on t(a,b,a) include(a,c,d,b);
+///
+/// Table t (called indexed table)
+/// columns:
+/// 0: `_row_id`
+/// 1: a
+/// 2: b
+/// 3: c
+/// 4: d
+///
+/// index idx (called index table)
+/// columns:
+/// 0: a
+/// 1: b
+/// 2: c
+/// 3: d
+/// 4: `t._row_id`
+///
+/// `TableDesc` columns of index table
+/// | index columns | include columns | hidden(pk columns) |
```

Comment:
This entire thing is the key, correct? As the pk column must be in the key to ensure uniqueness?

Reply:
The secondary index mv need to contain primary table pk to help later table lookup.
The key of secondary index is index_column + primary table pk.

Reply:
I found that we even support create an index on a materialized view, so we need to use order_key instead of pk.
The key of secondary index should be index_column + primary table order_key.

Reply:
Well, in that case, when we lookup the index, we'd have to do a range scan, correct?

Perhaps we can do only a point lookup if we encode lookup as:
key: index_cols
value: List<(pk, include_cols)>

?

Reply:
Yes, a range scan is inevitable for index lookup. For A reasonable index, the cardinality of the index column should be high,  so the cost of scan could be acceptable. We also want to create our index based on the method of creating materialized view and maybe expose our index as a table for users. If we encode lookup entries as you propose, it seems hard to express our index in a materialized view way.

```sql
create index idx on t(a) include(b)

equivalent to

create materialized view idx as select a, b from t order by a.
```

Reply:
> Perhaps we can do only a point lookup if we encode lookup entries as:
> 
> ```
> key: index_cols
> value: List<(pk, include_cols)>
> ```
> 
> ?

-1 for this.
we can do it with a new encoding of the relational table in the future. But currently, this encoding can not be expressed in our system and we must hack the relational storage interface to achieve that. 
this encoding is can change range scan to point look up, but as you say it has write amplification and might introduce a large size value in our kv storage. It is unacceptable if we have too many rows with the same index key.
this case is same as the stream join's state, I think we can use bloom filter to optimize the range scan.

Reply:
Ok, thanks for clarifying!